libuavcan_drivers/stm32/driver/src/uc_stm32_thread.cpp

```diff
@@ -79,6 +80,56 @@ void Mutex::unlock()
 # endif
 }
 
+
+#elif UAVCAN_STM32_FREERTOS
+
+bool BusEvent::wait(uavcan::MonotonicDuration duration)
+{
+    static const uavcan::int64_t MaxDelayMSec = 0x000FFFFF;
+
+    const uavcan::int64_t msec = duration.toMSec();
+
+    BaseType_t ret;
+
+    if (msec <= 0)
+    {
+        ret = xSemaphoreTake( sem_, ( TickType_t ) 0 );
+    }
+    else
+    {
+        ret = xSemaphoreTake( sem_, (msec > MaxDelayMSec) ? (MaxDelayMSec/portTICK_RATE_MS) : (msec/portTICK_RATE_MS));
+    }
+    return ret == pdTRUE;
+}
+
+void BusEvent::signal()
+{
+    xSemaphoreGive( sem_ );
+}
+
+void BusEvent::signalFromInterrupt()
+{
+    xHigherPriorityTaskWoken = pdFALSE;
+
+    xSemaphoreGiveFromISR( sem_, &xHigherPriorityTaskWoken );
+    
+    portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
```

Comment:
This seems to be a mis-use of the FreeRTOS API. Correct me if I'm wrong, but `portYIELD_FROM_ISR()` must be invoked only at the end of the ISR, which is not the case here - the driver may continue to perform other tasks in the ISR even after `signalFromInterrupt()` is invoked.